File: c/balanceddiet.c
// NAME : Balanced Diet
// URL  : https://open.kattis.com/problems/balanceddiet
// =============================================================================
// Use recursion to figure out all the possible sums we can encounter by
// combining values in all possible ways. Without any optimizations this
// solution is O(2<sup>N</sup>). However, the problem is conducive to dynamic
// programming since once a subproblem has been solved we can save the result
// leverage it to avoid redundant work.
// =============================================================================

#include <stdio.h>
#include <stdlib.h>

#define UNDEFINED -1
#define MAX_VALUES 20
#define MAX_MEMOIZE_VALUES 524288 // 14 bits for sum, 5 bits for n

struct result
{
    int sum1;
    int sum2;
    int diff;
};

struct result calc_min_diff(int* values, int n, int sum1, int total_sum, struct result* memoized)
{
    if (n == 0)
    {
        // BASE CASE: Return the absolute difference between the two sums.
        struct result r;
        r.sum1 = sum1;
        r.sum2 = total_sum - sum1;
        r.diff = abs(r.sum1 - r.sum2);

        return r;
    }

    // OPTIMIZATION: If we have seen this subproblem before, there is no sense doing the computation
    // again. We already know the result. Return it and save time. Given the integer constraints, we
    // can pack both values into a single int to serve as a fast dictionary key.
    int memoize_key = (n << 14) | sum1;
    struct result memoize_result = memoized[memoize_key];
    if (memoize_result.diff != UNDEFINED)
    {
        return memoize_result;
    }

    // RECURSION: From this point investigate both paths (including the item in the first
    // sum versus including it in the second sum).
    struct result a = calc_min_diff(values, n-1, sum1 + values[n-1], total_sum, memoized);
    struct result b = calc_min_diff(values, n-1, sum1, total_sum, memoized);

    // Best result = the one with the smallest difference between the sums.
    if (a.diff < b.diff)
    {
        memoized[memoize_key] = a;
        return a;
    }

    memoized[memoize_key] = b;
    return b;
}

struct result calc(int* values, int value_count)
{
    int total_sum = 0;
    for (int i = 0; i < value_count; i++)
    {
        total_sum += values[i];
    }

    // Initialize the memoize table by filling it with placeholder values.
    int memoized_len = MAX_MEMOIZE_VALUES * sizeof(struct result);
    struct result* memoized = malloc(memoized_len);
    memset(memoized, UNDEFINED, memoized_len);

    // Kick off the recursion 
    struct result r = calc_min_diff(values, value_count, 0, total_sum, memoized);
    free(memoized);

    return r;
}


int main()
{
    int value_count;
    scanf("%d", &value_count);

    while (value_count != 0)
    {
        int values[MAX_VALUES];
        for (int i = 0; i < value_count; i++)
        {
            scanf("%d", &values[i]);
        }

        struct result r = calc(values, value_count);

        // Make sure the larger value is put first.
        if (r.sum1 > r.sum2)
        {
            printf("%d %d\n", r.sum1, r.sum2);
        }
        else
        {
            printf("%d %d\n", r.sum2, r.sum1);
        }

        scanf("%d", &value_count);        
    }

    return 0;
}

File: cs/balanceddiet.cs
// NAME : Balanced Diet
// URL  : https://open.kattis.com/problems/balanceddiet
// =============================================================================
// Use recursion to figure out all the possible sums we can encounter by
// combining values in all possible ways. Without any optimizations this
// solution is O(2<sup>N</sup>). However, the problem is conducive to dynamic
// programming since once a subproblem has been solved we can save the result
// leverage it to avoid redundant work. Solved in C# to compare language
// performance versus C on judging machines.
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

public class Program
{
    public static void Main()
    {
        string line = Console.ReadLine();
        while (line != "0")
        {
            // Ignore the count and just read the remaining numbers into an array (length comes for
            // free via the array)
            int[] values = line.Split().Skip(1).Select(int.Parse).ToArray();
            Result result = CalcMinDiff(values);

            // Make sure the larger value is put first.
            if (result.Sum1 > result.Sum2)
            {
                Console.WriteLine("{0} {1}", result.Sum1, result.Sum2);
            }
            else
            {
                 Console.WriteLine("{0} {1}", result.Sum2, result.Sum1);
            }

            line = Console.ReadLine();
        }
    }

    public static Result CalcMinDiff(int[] arr)
    {
        var memoizedSums = new Dictionary<int, Result>();
        int totalSum = arr.Sum();

        // Kick off the recursion 
        return CalcMinDiff(arr, arr.Length, 0, totalSum, memoizedSums);
    }

    public static Result CalcMinDiff(int[] arr, int n, int sum1, int totalSum, Dictionary<int, Result> memoizedSums)
    {
        if (n == 0)
        {
            int sum2 = totalSum - sum1;

            // BASE CASE: Return the absolute difference between the two sums.
            return new Result { Sum1 = sum1, Sum2 = sum2, Diff = Math.Abs(sum1 - sum2) };
        }

        // OPTIMIZATION: If we have seen this subproblem before, there is no sense doing the computation
        // again. We already know the result. Return it and save time. Given the integer constraints, we
        // can pack both values into a single int to serve as a fast dictionary key.
        int memoizeKey = (n << 16) | sum1;
        Result memoizeResult;
        if (memoizedSums.TryGetValue(memoizeKey, out memoizeResult))
        {
            return memoizeResult;
        }

        // RECURSION: From this point investigate both paths (including the item in the first
        // sum versus including it in the second sum).
        Result resultA = CalcMinDiff(arr, n-1, sum1 + arr[n-1], totalSum, memoizedSums);
        Result resultB = CalcMinDiff(arr, n-1, sum1, totalSum, memoizedSums);

        // Best result = the one with the smallest difference between the sums.
        Result best = resultA.Diff < resultB.Diff
            ? resultA
            : resultB;

        memoizedSums.Add(memoizeKey, best);

        return best;
    }

    public struct Result
    {
        public int Sum1;
        public int Sum2;
        public int Diff;
    }
}